### PR TITLE
[flutter_tools] delegate first run message re-display to new class, only if changed

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Fixes https://github.com/flutter/flutter/issues/assigned/jonahwilliams, by computing/caching a hash of the first run message and only display it when it changes. This will prevent it from being printed outside of changes to the message or first run scenarios.

### CODE REVIEW

**Review by zanderso — APPROVED** [review]
lgtm

File: packages/flutter_tools/lib/src/reporting/usage.dart [hunk 1]
```diff
@@ -383,52 +388,19 @@ class _DefaultUsage implements Usage {
...
-  }
-
   @override
   void printWelcome() {
     // Only print once per run.
     if (_printedWelcome) {
       return;
     }
-    if (// Display the welcome message if this is the first run of the tool.
-        isFirstRun ||
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/reporting/usage.dart:425]: Nope, removed

File: packages/flutter_tools/lib/src/reporting/first_run.dart [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+  FirstRunMessenger({
+    @required PersistentToolState persistentToolState
+  }) : _persistentToolState = persistentToolState;
+
+  final PersistentToolState _persistentToolState;
+
+  /// Whether the license terms should be displayed.
+  ///
+  /// This is implemented by caching a hash of the previous license terms. This
+  /// does not updated the cache hash value.
```
Comment (MrBirb) [inline packages/flutter_tools/lib/src/reporting/first_run.dart:55]: nit: update*

File: packages/flutter_tools/lib/src/reporting/first_run.dart [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  ///
+  /// The persistent tool state setting [PersistentToolState.redisplayWelcomeMessage]
+  /// can also be used to make this return false. This is primarily used to ensure
+  /// that the license terms are not printed during a `flutter upgrade`, until the
+  /// use manually runs the tool.
```
[the remark this answers is not in the document]
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/reporting/first_run.dart]: Fixed

File: packages/flutter_tools/lib/src/reporting/first_run.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/reporting/first_run.dart]: Fixed

**Comment by jonahwilliams** [comment]
Thanks @MrBirb !

**Pull request merged**